src/webgpu/api/validation/layout_shader_compat.spec.ts

```diff
@@ -1,14 +1,318 @@
+  'compareSamp',
+  'sampledTex',
+  'sampledTexMS',
+  'readonlyStorageTex',
+  'writeonlyStorageTex',
+  'readwriteStorageTex',
+] as const;
+
+class F extends ValidationTest {
+  createPipelineLayout(
+    bindingInPipelineLayout: BindableResourceType,
+    visibility: number
+  ): GPUPipelineLayout {
+    let bindGroupLayoutEntry: GPUBindGroupLayoutEntry;
+    switch (bindingInPipelineLayout) {
+      case 'compareSamp': {
+        bindGroupLayoutEntry = {
+          binding: 0,
+          visibility,
+          sampler: {
+            type: 'comparison',
+          },
+        };
+        break;
+      }
+      case 'filtSamp': {
+        bindGroupLayoutEntry = {
+          binding: 0,
+          visibility,
+          sampler: {
+            type: 'filtering',
+          },
+        };
+        break;
+      }
+      case 'nonFiltSamp': {
+        bindGroupLayoutEntry = {
+          binding: 0,
+          visibility,
+          sampler: {
+            type: 'non-filtering',
+          },
+        };
+        break;
+      }
+      case 'sampledTex': {
+        bindGroupLayoutEntry = {
+          binding: 0,
+          visibility,
+          texture: {
+            sampleType: 'unfilterable-float',
+          },
+        };
+        break;
+      }
+      case 'sampledTexMS': {
+        bindGroupLayoutEntry = {
+          binding: 0,
+          visibility,
+          texture: {
+            sampleType: 'unfilterable-float',
+            multisampled: true,
+          },
+        };
+        break;
+      }
+      case 'storageBuf': {
+        bindGroupLayoutEntry = {
+          binding: 0,
+          visibility,
+          buffer: {
+            type: 'storage',
+          },
+        };
+        break;
+      }
+      case 'readonlyStorageBuf': {
+        bindGroupLayoutEntry = {
+          binding: 0,
+          visibility,
+          buffer: {
+            type: 'read-only-storage',
+          },
+        };
+        break;
+      }
+      case 'uniformBuf': {
+        bindGroupLayoutEntry = {
+          binding: 0,
+          visibility,
+          buffer: {
+            type: 'uniform',
+          },
+        };
+        break;
+      }
+      case 'readonlyStorageTex': {
+        bindGroupLayoutEntry = {
+          binding: 0,
+          visibility,
+          storageTexture: {
+            format: 'r32float',
+            access: 'read-only',
+          },
+        };
+        break;
+      }
+      case 'writeonlyStorageTex': {
+        bindGroupLayoutEntry = {
+          binding: 0,
+          visibility,
+          storageTexture: {
+            format: 'r32float',
+            access: 'write-only',
+          },
+        };
+        break;
+      }
+      case 'readwriteStorageTex': {
+        bindGroupLayoutEntry = {
+          binding: 0,
+          visibility,
+          storageTexture: {
+            format: 'r32float',
+            access: 'read-write',
+          },
+        };
+        break;
+      }
+    }
```

Comment:
Or maybe use a dictionary (e.g. `bindGroupLayoutEntryContents`) to map `bindingInPipelineLayout` to the content (`buffer`, `sampler` and others), and we can have a simpler
``` typescript
    const bindGroupLayoutEntryContents = {
      'compareSamp': {
        sampler: {
          type: 'comparison',
        },
      },
      'filtSamp': ...,
      ...
    } as const;

    bindGroupLayoutEntry = {
      binding: 0,
      visibility,
      ...bindGroupLayoutEntryContents[bindingInPipelineLayout],
    };
```